### PR TITLE
Removed cancellation of all outstanding requests in discovery service

### PR DESCRIPTION
**Pull request by kopaygorodsky**
I had a case when I was asking for membership on a channel from 3 peers. Only 1 was joined to a channel. Targets are randomly selected and if first peer wasn't the right one I get response value "access denied" which was added to responses array and counted as "successful". Then context is canceled and request never has been sent to the right peer. 

So basically I think we need to remove ctx cancelation at that point  because message evaluation is done on lover levels and we can't ignore responses from other peers. It won't affect performance a lot.

### CODE REVIEW

**Comment by kopaygorodsky** [comment]
@bstasyszyn could you check it pls, I see you were working on it

**Review by bstasyszyn — COMMENTED** [review]
The initial reason that cancel was added because we had cases where a request was hanging on a bad connection and the client had to wait for the timeout even though it already had a response from another peer. A better strategy would be to cancel only if we already have a -successful- response.

**Comment by kopaygorodsky** [comment]
@bstasyszyn it means that we need to evaluate message at that point? because value "Access denied" is valid right now.

**Comment by bstasyszyn** [comment]
> @bstasyszyn it means that we need to evaluate message at that point? because value "Access denied" is valid right now.

Yes, the code would need to be refactored a bit. On first glance, I would add a 'validator' function to be passed in as an arg to DiscoveryClient.Send. Invoke the validator func on each response and, if no error, cancel and return.

**Comment by kopaygorodsky** [comment]
@bstasyszyn Canceling request under the hood is not ok, because users expect that request will go to all targets he specified. If some peer is hanging - it's not DS Client's problem, there is ctx and timeout for it. 
I propose to change Send signature in favour of returning `chan Response`. Then we will need to fix only one component that uses it - chservice.go. 

```
	for resp := range responses {
		peer, err := s.evaluate(ctx, resp)

		if err == nil {
			//Cancel all outstanding requests
			cancel()
		}
		
		// do some work
	}
```

**Comment by bstasyszyn** [comment]
> @bstasyszyn Canceling request under the hood is not ok, because users expect that request will go to all targets he specified. If some peer is hanging - it's not DS Client's problem, there is ctx and timeout for it.
> I propose to change Send signature in favour of returning `chan Response`. Then we will need to fix only one component that uses it - chservice.go.
> 
> ```
> 	for resp := range responses {
> 		peer, err := s.evaluate(ctx, resp)
> 
> 		if err == nil {
> 			//Cancel all outstanding requests
> 			cancel()
> 		}
> 		
> 		// do some work
> 	}
> ```

Yes, this looks good.

**Comment by kopaygorodsky** [comment]
@bstasyszyn what is the reason of requesting a single target for local peers? 
https://github.com/hyperledger/fabric-sdk-go/blob/master/pkg/client/common/discovery/dynamicdiscovery/localservice.go#L80

Can we request all known peers from endpoint config with mspID from ctx?

**Comment by kopaygorodsky** [comment]
1. Moved DisoveryClient interface to pkg/fab/discovery + mocks, all users can rely on this interface.
2. Reworked pkg/fab/discovery Send, now it returns a channel of Responses, cancellation of the context is delegated to users of this client. Error method was added to Response, so user can check if an error happened.
3. Fixed tests
4. Reworked chservice, localservice, fabricselection by using chan Response
5. Updated DiscoveryError by adding IsAccessDenied method, no need to expose const and compare msg every time.

**Comment by kopaygorodsky** [comment]
All tests passed except integration one`TestQueryCollectionsConfig`, maybe it has something to do with local discovery because I changed to requesting all peers from endpoint config.  I can change it back once you check on it. Need some guidance to move forward with this, there are plenty of changes.

File: pkg/fab/discovery/discovery.go [hunk 1]
```diff
@@ -29,20 +28,27 @@ const (
 	signerCacheSize = 10 // TODO: set an appropriate value (and perhaps make configurable)
 )
 
+//Client gives ability to send discovery request to multiple targets.
+//There are cases when multiple targets requested and some of them are hanging, recommended to cancel ctx after first successful response.
+//Note: "access denied" is a success response, so check for it after response evaluation.
+type Client interface {
```
Comment (bstasyszyn) [inline pkg/fab/discovery/discovery.go]: I wouldn't define the interface here. In Go, interfaces are (usually) defined at the receiving end. The Client struct doesn't make use of this interface at all. If you wish to define it in a common place then it should be defined in pkg/client/common.

File: pkg/fab/discovery/discovery.go [hunk 2]
```diff
@@ -29,20 +28,27 @@ const (
...
 // Client implements a Discovery client
-type Client struct {
+type client struct {
 	ctx      fabcontext.Client
 	authInfo *discovery.AuthInfo
 }
 
 // New returns a new Discover client
-func New(ctx fabcontext.Client) (*Client, error) {
+func New(ctx fabcontext.Client) (Client, error) {
```
Comment (bstasyszyn) [inline pkg/fab/discovery/discovery.go]: This function should return a pointer to a struct and not an interface (according to Go conventions).

**Review by bstasyszyn — COMMENTED** [review]
This change looks good. I just left a couple of comments.

File: pkg/fab/discovery/discovery.go (hunk 2 above)
Comment (kopaygorodsky) [reply pkg/fab/discovery/discovery.go]: I was following https://golang.org/doc/effective_go.html#generality. 

Good, even I'm not a fun of `accept interfaces, return structs` :)

Comment (bstasyszyn) [reply pkg/fab/discovery/discovery.go]: > I was following https://golang.org/doc/effective_go.html#generality.
> 
> Good, even I'm not a fun of `accept interfaces, return structs` :)

Hmm - I guess I learned something new today :)

**Review by bstasyszyn — APPROVED** [review]

File: pkg/fab/discovery/discovery.go (hunk 1 above)
Comment (kopaygorodsky) [reply pkg/fab/discovery/discovery.go]: I'm considering discovery service as a component(like net.Dialer), so coupling on this package is ok.

What if I make Client struct as a private one? no reason to expose it, no functionality beyond interface.

**Comment by troyronda** [comment]
@kopaygorodsky The test coverage is low: 66%

**Comment by kopaygorodsky** [comment]
@troyronda I delivered 0 of new functionality :)  
heh, I'll try to increase it

**Comment by kopaygorodsky** [comment]
@troyronda is it because this functionality wasn't covered before ?  absolute coverage is 74%, but relative is 67%. What parts should I cover more? Everything I was working on is tested by units. 
1) discovery client (Send) 
2) chservice GetPeers is tested
3) localprodiver GetPeers is tested
4) fabricselection GetPeers too

**Comment by kopaygorodsky** [comment]
I think there is smth wrong with https://codecov.io/gh/hyperledger/fabric-sdk-go/compare/48e23299095fd200363d8a14edf56b4e14c5973b...a518e1a76a4f2db06c75fdfd71971c56d7066b27/src/pkg/fab/discovery/mockdiscoveryclient.go. I think we can add mocks to ignore list,

**Comment by kopaygorodsky** [comment]
Guys, I've covered everything related to my PR as much as possible, if you check report - everything is green. One commit before it was 70%, I added a few tests to cover error handling and now it's 69%...
Can we just merge it?  my PR brings +0.15% to the table of total coverage. 74.21 -> 74.35
I think, it's fine, no?

File: pkg/client/common/discovery/dynamicdiscovery/localprovider_test.go [hunk 3]
```diff
@@ -9,10 +9,10 @@ SPDX-License-Identifier: Apache-2.0
 package dynamicdiscovery
 
 import (
+	"github.com/hyperledger/fabric-sdk-go/pkg/fab/discovery"
```
Comment (troyronda) [inline pkg/client/common/discovery/dynamicdiscovery/localprovider_test.go:12]: import should be in second group.

File: pkg/client/common/discovery/dynamicdiscovery/localservice_test.go [hunk 4]
```diff
@@ -9,14 +9,14 @@ SPDX-License-Identifier: Apache-2.0
 package dynamicdiscovery
 
 import (
-	"errors"
+	"github.com/pkg/errors"
```
Comment (troyronda) [inline pkg/client/common/discovery/dynamicdiscovery/localservice_test.go:12]: import shouldn't be in stdlibrary imports group.

File: pkg/client/common/selection/fabricselection/selection_test.go [hunk 5]
```diff
@@ -10,11 +10,12 @@ package fabricselection
 
 import (
 	"fmt"
+	"github.com/hyperledger/fabric-sdk-go/pkg/fab/discovery"
```
Comment (troyronda) [inline pkg/client/common/selection/fabricselection/selection_test.go:13]: imports are in wrong group.

File: pkg/client/common/selection/fabricselection/selection_test.go [hunk 6]
```diff
@@ -210,16 +211,90 @@ func TestSelection(t *testing.T) {
...
+	serviceNoErrHandling, err := New(
+		ctx, channelID,
+		mocks.NewMockDiscoveryService(nil, peer1Org1, peer2Org1, peer1Org2, peer2Org2, peer1Org3, peer2Org3),
+		WithRefreshInterval(5*time.Millisecond),
+		WithResponseTimeout(100*time.Millisecond),
+	)
+	require.NoError(t, err)
+	defer serviceNoErrHandling.Close()
+
+	//this is incorrect way to test behaviour when access denied. The message should be propagated inside evaluating func, only at this step we can get such err
```
Comment (troyronda) [inline pkg/client/common/selection/fabricselection/selection_test.go:223]: left over code?

**Review by troyronda — APPROVED** [review]

**Comment by kopaygorodsky** [comment]
mockdiscoveryclient.go is still counted.

File: pkg/client/common/selection/fabricselection/selection_test.go (hunk 5 above)
Comment (kopaygorodsky) [reply pkg/client/common/selection/fabricselection/selection_test.go]: fixed. How did it pass linter?

**Comment by kopaygorodsky** [comment]
Integration tests failed for some unknown reason, were succeed a few commits before and nothing changes in terms of functionality.

**Comment by kopaygorodsky** [comment]
it fixed itself somehow, concurrency problems, maybe. 
Ready to be merged. @troyronda

**Pull request merged**